sdks/python/apache_beam/runners/interactive/interactive_beam.py

```diff
@@ -0,0 +1,199 @@
+  scope to instruct iBeam to apply magic to your pipeline when running pipeline
+  interactively.
+
+    For example:
+
+    class Foo(object)
+      def build_pipeline(self):
+        p = create_pipeline()
+        init_pcoll = p |  'Init Create' >> beam.Create(range(10))
+        watch(locals())
+        return p
+    Foo().build_pipeline().run()
+
+    iBeam will cache init_pcoll for the first run. You can use:
+
+    visualize(init_pcoll)
+
+    To visualize data from init_pcoll once the pipeline is executed. And if you
+    make change to the original pipeline by adding:
+
+    squares = init_pcoll | 'Square' >> beam.Map(lambda x: x*x)
+
+    When you re-run the pipeline from the line you just added, squares will
+    use the init_pcoll data cached so you can have an interactive experience.
+
+  Currently the implementation mainly watches for PCollection variables defined
+  in user code. A watchable can be a dictionary of variable metadata such as
+  locals(), a str name of a module, a module object or an instance of a class.
+  The variable can come from any scope even local variables in a method of a
+  class defined in a module.
+
+    Below are all valid:
+
+    watch(__main__)  # if import __main__ is already invoked
+    watch('__main__')  # does not require invoking import __main__ beforehand
+    watch(self)  # inside a class
+    watch(SomeInstance())  # an instance of a class
+    watch(locals())  # inside a function, watching local variables within
+  """
+  current_env().watch(watchable)
+
+
+def create_pipeline(runner=None, options=None, argv=None):
+  """Creates a pipeline with interactive runner by default.
+
+  You can use run_pipeline() provided within this module to execute the iBeam
+  pipeline with other runners.
+
+  Args:
+    runner (~apache_beam.runners.runner.PipelineRunner): An object of
+      type :class:`~apache_beam.runners.runner.PipelineRunner` that will be
+      used to execute the pipeline. For registered runners, the runner name
+      can be specified, otherwise a runner object must be supplied.
+    options (~apache_beam.options.pipeline_options.PipelineOptions):
+      A configured
+      :class:`~apache_beam.options.pipeline_options.PipelineOptions` object
+      containing arguments that should be used for running the Beam job.
+    argv (List[str]): a list of arguments (such as :data:`sys.argv`)
+      to be used for building a
+      :class:`~apache_beam.options.pipeline_options.PipelineOptions` object.
+      This will only be used if argument **options** is :data:`None`.
+
+  Raises:
+    ~exceptions.ValueError: if either the runner or options argument is not
+      of the expected type.
+  """
+  if not runner:
+    runner = interactive_runner.InteractiveRunner()
+  return beam.Pipeline(runner, options, argv)
+
+
+def visualize(pcoll):
+  """Visualizes a PCollection."""
+  # TODO(ningk)
+  pass
+
+
+def run_pipeline(pipeline, runner=None, options=None):
```

Comment:
What would this do? Since create_pipeline above will set the runner and options already for the pipeline.

Reply:
The run_pipeline takes in an existing interactive pipeline object, a new runner and options for that new runner so that user doesn't need to change/modify/re-execute any code in notebook to run the pipeline on selected platform say Dataflow by adding a new notebook cell and execute the single line of code.

We'll just have a variation of run_with_dataflow() at this time. I can replace this with run_with_dataflow() if it resolves confusion.